### PR TITLE
[GPU] MaxPool-8

### PR DESCRIPTION
**Pull request by okhovan**
- added MaxPool-8 features (dialtions and 2nd output for indices)
    - MaxPool-1 behaviour is preserved.

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/ops/pooling.cpp [hunk 1]
```diff
@@ -97,7 +113,57 @@ static void CreateMaxPoolOp(Program& p, const std::shared_ptr<ngraph::op::v1::Ma
...
+
+    auto mutable_precision = op->get_output_element_type(1);
+    cldnn::layout mutableLayout = cldnn::layout(DataTypeFromPrecision(mutable_precision),
+                                                DefaultFormatForDims(op->get_output_shape(1).size()),
+                                                CldnnTensorFromIEDims(op->get_output_shape(1)));
+    auto shared_memory = p.GetEngine().allocate_memory(mutableLayout);
+    cldnn::primitive_id maxpool_mutable_id_w = layer_type_name_ID(op) + "_md_write";
+    auto indices_mutable_prim = cldnn::mutable_data(maxpool_mutable_id_w,
+                                                   shared_memory,
+                                                   op->get_friendly_name());
```
Comment (serhii-pavlovskyi-altran) [inline src/plugins/intel_gpu/src/plugin/ops/pooling.cpp:133]: `op->get_friendly_name()` potentially could be non-trivial and is called 3 times here

File: src/plugins/intel_gpu/src/plugin/ops/pooling.cpp [hunk 2]
```diff
@@ -97,7 +113,57 @@ static void CreateMaxPoolOp(Program& p, const std::shared_ptr<ngraph::op::v1::Ma
...
+    p.primitiveIDs[maxpool_mutable_id_w] = maxpool_mutable_id_w;
+    p.AddPrimitive(indices_mutable_prim);
+    inputPrimitives.push_back(maxpool_mutable_id_w);
+
+    const auto params = GetPoolingParameters(op->get_kernel(), op->get_strides(), op->get_pads_begin(), op->get_pads_end(), op->get_dilations());
+    auto poolPrim = cldnn::pooling(layerName,
+                                   inputPrimitives[0],
+                                   inputPrimitives[inputPrimitives.size() - 1],
```
Comment (serhii-pavlovskyi-altran) [inline src/plugins/intel_gpu/src/plugin/ops/pooling.cpp:141]: `inputPrimitives.back()`

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/pooling/pooling_kernel_base.cpp [hunk 3]
```diff
@@ -60,6 +60,26 @@ JitConstants PoolingKernelBase::GetJitConstants(const pooling_params& pp, Poolin
...
+    if (pp.maxPoolOpset8Features) {
+        mem_consts.AddConstants({
+            MakeJitConstant("DILATION", pp.poolDilation),
+            MakeJitConstant("AXIS", pp.poolAxis)
+        });
+
+        if (pp.poolAxis != 0) {
+            size_t indices_upper_bound = 1;
+            const auto& dims = pp.inputs[0].GetDims();
+            for (auto d = dims.crbegin() + pp.poolAxis; d != dims.crend(); ++d) {
```
Comment (serhii-pavlovskyi-altran) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/pooling/pooling_kernel_base.cpp:69]: this loop could be replaced with `std::accumulate`

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl [hunk 4]
```diff
@@ -116,20 +123,34 @@ KERNEL(pooling_gpu)(
 #if OUTPUT_DIMS == 5
     for(uint l = 0; l < POOL_SIZE_Z; l++)
     {
-        int input_offset_z = offset_z + l;
+        #if MAX_POOLING && defined(DILATION_SIZE_Z)
+            int input_offset_z = offset_z + (l * DILATION_SIZE_Z);
```
Comment (serhii-pavlovskyi-altran) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl:127]: if you define `DILATION_SIZE_*` to `1` when they are not used, you can get rid of `#ifdef` dances here, compiler will optimize away multiplication by `1`

File: inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/pooling.hpp [hunk 5]
```diff
@@ -89,17 +89,6 @@ typedef std::tuple<
...
-typedef std::tuple<
-        maxPool8SpecificParams,
-        InferenceEngine::Precision,     // Net precision
-        InferenceEngine::Precision,     // Input precision
-        InferenceEngine::Precision,     // Output precision
-        InferenceEngine::Layout,        // Input layout
-        InferenceEngine::Layout,        // Output layout
-        size_t,                         // Channel number
-        std::string                     // Device name
-> globalMaxPool8LayerTestParamsSet;
```
Comment (serhii-pavlovskyi-altran) [inline inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/pooling.hpp]: this shouldn't be added in first commit

File: inference-engine/thirdparty/clDNN/api/cldnn/primitives/pooling.hpp [hunk 6]
```diff
@@ -204,7 +204,7 @@ struct pooling : public primitive_base<pooling> {
               pad_end(pad_end),
               axis(axis),
               index_element_type(index_element_type),
-              maxPoolOpset8Features{true}
+              maxPoolOpset8Features(true)
```
Comment (serhii-pavlovskyi-altran) [inline inference-engine/thirdparty/clDNN/api/cldnn/primitives/pooling.hpp:207]: i'd stick to recommended practice even if surrounding code doesn't

File: inference-engine/tests/functional/shared_test_classes/src/single_layer/pooling.cpp [hunk 7]
```diff
@@ -107,7 +107,7 @@ std::string MaxPool8LayerTest::getTestCaseName(const testing::TestParamInfo<maxP
     ngraph::op::PadType padType;
     ngraph::op::RoundingType roundingType;
     ngraph::element::Type indexElementType;
-    int32_t axis;
+    int64_t axis;
```
Comment (serhii-pavlovskyi-altran) [inline inference-engine/tests/functional/shared_test_classes/src/single_layer/pooling.cpp:110]: you should fix earlier commits with `git commit --fixup` + `git rebase`, changing same piece of code back and forth complicates review and git history. same for test of changes in this commit

**Review by serhii-pavlovskyi-altran — CHANGES_REQUESTED** [review]

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl [hunk 8]
```diff
@@ -159,7 +180,17 @@ KERNEL(pooling_gpu)(
...
+                            #ifdef SECOND_OUTPUT_TYPE
+                                if (casted_input > result)
+                                {
+                                    result = casted_input;
+                                    result_idx = input_idx;
+                                }
+                            #else
+                                result = FUNC_CALL(apply_pooling)(result, casted_input);
+                            #endif
+
```
Comment (ytorzuk-altran) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl:193]: MaxPool-8 should be implemented for `CHECK_BOUNDARY == 0` case as well. You can add a test for MaxPool-8 where feature map spatial dimensions are evenly divisible by kernel size to validate this case.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl [hunk 9]
```diff
@@ -227,7 +258,15 @@ KERNEL(pooling_gpu)(
                 uint input_idx = INPUT0_GET_INDEX(b, f, offset_y + j, offset_x + i);
                 result = FUNC_CALL(apply_pooling)(result, TO_ACCUMULATOR_TYPE(input[input_idx]));
     #else
-                result = FUNC_CALL(apply_pooling)(result, TO_ACCUMULATOR_TYPE(input[input_idx]));
+                #ifdef SECOND_OUTPUT_TYPE
+                    if (input[input_idx] > result)
```
Comment (ytorzuk-altran) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl:262]: Minimize reading from the global memory. One of solutions would be to have an additional variable which keeps previous result value.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl [hunk 10]
```diff
@@ -28,6 +28,9 @@ KERNEL(pooling_gpu)(
 #if MAX_WITH_ARGMAX_POOLING
 , __global float* arg_max
 #endif
+#ifdef SECOND_OUTPUT_TYPE
```
Comment (ytorzuk-altran) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_ref.cl:31]: A dedicated macro (like `OUTPUT_SELECTED_INDICES`) would improve readability

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/pooling/pooling_kernel_gpu_ref.cpp [hunk 11]
```diff
@@ -27,6 +27,7 @@ ParamsKey PoolingKernelGPURef::GetSupportedKey() const {
     k.EnablePoolKernelDividerMode(KernelDividerMode::DYNAMIC);
     k.EnablePoolKernelDividerMode(KernelDividerMode::DYNAMIC_WITH_PADDING);
     k.EnableDifferentTypes();
+    k.EnableMaxPoolOpset8Features();
```
Comment (ytorzuk-altran) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/pooling/pooling_kernel_gpu_ref.cpp:30]: MaxPool-8 has 2 additional features comparing to MaxPool-1. EnableMaxPoolOpset8Features key can be split into 2 keys:
* EnableDilations
* EnableOutputSelectedIndeces

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/pooling/pooling_kernel_base.cpp [hunk 12]
```diff
@@ -60,6 +60,23 @@ JitConstants PoolingKernelBase::GetJitConstants(const pooling_params& pp, Poolin
...
+            if (indices_upper_bound != 0 && indices_upper_bound != 1) {
+                mem_consts.AddConstants({MakeJitConstant("INDICES_UPPER_BOUND", indices_upper_bound)});
+            }
+        }
+
+        mem_consts.Merge(MakeTypeJitConstants(pp.poolIndexElementType, "SECOND_OUTPUT"));
+    }
+
     if (dispatchData.needsBoundary) {
         mem_consts.AddConstant(MakeJitConstant("CHECK_BOUNDRY", 1));
```
Comment (ytorzuk-altran) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/pooling/pooling_kernel_base.cpp:81]: Spelling issue: CHECK_BOUND**A**RY

**Pull request closed without merging**